lldb: get lldb API tests working with newer Android NDKs

andrurogerz opened the pull request:
## Purpose
Running the LLDB API tests against a remote Android target with NDK version r22 or later fails to compile the test inferiors. NDK r21 from 2021 is the most recent NDK that still works with the LLDB API tests. This PR updates the Android make rules to support newer Android NDK versions (r19 and later).

## Overview
* Updates and simplifies `Android.rules` to match the newer Android NDK unified toolchain layout introduced in NDK r19
* Sets `OBJCOPY` and `ARCHIVER` env vars, required by a few test cases, to their `llvm-` versions in the unified toolchain
* Drops support for pre-2019 Android NDK versions to keep the rules simple
* Provides an error message if the tests are run using an incompatible NDK layout

## Problem Details
Android introduced a unified tools layout in NDK r19 (2019) and removed the old layout in r22 (2021). Releases r19, r20, and r21 support both the old and new layout side-by-side. More details are in #106270.

## Validation
Ran a sub-set of the LLDB API tests against remote Android targets for the four primary architectures i386, x86_64, arm, and aarch64. No validation was done against riscv targets.

For each case, ran the copy of `lldb-server` from the Android NDK on the device with the latest LLDB test cases in llvm-project

Ran tests with both r19 (the oldest supported) and r26 (more recent, unified layout only) NDK versions. 

Example test command for aarch64:
```
./build/bin/lldb-dotest --out-of-tree-debugserver --arch aarch64 --platform-name remote-android --platform-url connect://localhost:5432 --platform-working-dir /data/local/tmp --compiler=$ANDROID_NDK_ROOT/toolchains/llvm/prebuilt/linux-x86_64/bin/clang lldb/test/API/android/
```
**NOTE: there are a lot of test failures when running the full suite (especially against 32-bit ARM target). These failures occur independent of this change.**

Verified the expected error message appears when attempting to run using NDK r18
```
Build Command Output:
make: Entering directory '/home/andrew/src/llvm/llvm-project/build/lldb-test-build.noindex/android/platform/TestDefaultCacheLineSize.test_cache_line_size'
/home/andrew/src/llvm/llvm-project/lldb/packages/Python/lldbsuite/test/make/Android.rules:16: *** "No unified toolchain sysroot found in /home/andrew/Android/Sdk/ndk/18.1.5063045/toolchains/llvm/prebuilt/linux-x86_64/bin/../../../../... NDK must be r19 or later.".  Stop.
make: Leaving directory '/home/andrew/src/llvm/llvm-project/build/lldb-test-build.noindex/android/platform/TestDefaultCacheLineSize.test_cache_line_size'
```

## Impact
**This change explicitly removes support for the pre-2019 NDK structure.** Only NDK r19 (from 2019) and later can be used when running the LLDB API tests.  If the maintainers object, we can easily support both the old and new NDK toolchain layouts side-by-side at the cost of readability/maintainability. Since this change only impacts tests, I don't see much value in supporting NDKs that are over 5 years old.

## Guidance to Reviewers
* I am not an expert on `clang` arguments so if anything looks off let me know.
* While I personally thing supporting 5+ year old NDKs for testing seems unnecessary, please chime-in if you are concerned with dropping that support. I can easily revise to support both old and new layouts side-by-side.
* If there are any specific tests you'd like me to run I will do my best to accommodate. It doesn't look like there's much (any?) Android LLDB CI coverage.

andrurogerz committed:
[android] get lldb tests working against newer Android NDKs

Android introduced a unified tools layout in NDK r19 (2019) and
removed the old layout in r22 (2021). Running lldb tests with NDK r22
or newer fails when compiling the test inferiors.

This change updates `Android.rules` to match the newer unified tools
structure introduced in NDK r19, which is described in more detail at
https://github.com/android/ndk/issues/780.

NOTE: After this change, ONLY NDK r19c and later can be used when
running the lldb tests. The pre-2019 NDK structure is no longer
supported.

andrurogerz committed:
[android] explicitly fail inferior compilation on old NDKs

Provide an error message when compiling LLDB API test inferiors against
an older Android NDK without the unified toolchain layout. This message
will help guide developers to install and use an NDK r19 or newer.

labath commented on lldb/packages/Python/lldbsuite/test/make/Android.rules:
```diff
@@ -1,81 +1,55 @@
+	--target=$(TRIPLE) \
+	--sysroot=$(TOOLCHAIN_SYSROOT) \
+	--prefix=$(TOOLCHAIN_SYSROOT)/usr/lib/$(ARCH_DIR)/$(API_LEVEL) \
+	--library-directory=$(TOOLCHAIN_SYSROOT)/usr/lib/$(ARCH_DIR)/$(API_LEVEL) \
```
```suggestion
	-L$(TOOLCHAIN_SYSROOT)/usr/lib/$(ARCH_DIR)/$(API_LEVEL) \
```

I haven't seen anyone use the long version of this option (I had to look it up to be sure of what it does), so I think it'd be better to stick to `-L`. `--prefix` is probably ok, since its short form (`-B`) is also not very common.

labath commented on lldb/packages/Python/lldbsuite/test/make/Android.rules:
```diff
@@ -1,81 +1,55 @@
+	-lm \
 	-lc++abi \
-	-nostdlib++
+	-nostdlib++ \
```
I get the diff benefits of this, but a backslash continuation at the end of a file looks very suspicious. I'd just remove all of the final backslashes you have here.

labath reviewed (commented):
This looks great. I don't think anyone will object to removing old ndk support, but let's leave this PR open for a while to give people a chance to notice.

andrurogerz committed:
Removing final trailing \ and switch back to -L

andrurogerz committed:
[android] use llvm-objcopy and -ar from NDK location

The NDK unified toolchain layout includes `llvm-objcopy` and `llvm-ar`
host executables. A few of the lldb API tests rely on these tools.
Previously, the GCC verions of these tools were used from their old NDK
location which no longer exists in post-r22 NDKs. Instead, use the llvm
versions of these tools which are included in the unified toolchain.

andrurogerz commented:
I applied feedback from @labath (thank you!) and am happy to leave this PR up for a while in case anyone else has thoughts.

I also made one improvement:

It now sets `OBJCOPY=` and `ARCHIVER=` to `llvm-objcopy` and `llvm-ar` included since NDK r19. These vars were previously pointed at the binutils versions that were removed in NDK r22. A small number of the tests rely on `objcopy` and `ar`, so explicitly pointing these vars at the `llvm-` variants in the NDK avoids requiring them be passed to `lldb-dotest` via `--env` command line args.

andrurogerz commented:
@labath there have been no additional comments; is there anyone specifically you'd like me to seek feedback from?

@JDevlieghere would you have a moment to take a look at this PR soon? I really appreciate it.

@compnerd FYI

compnerd commented on lldb/packages/Python/lldbsuite/test/make/Android.rules:
```diff
@@ -1,81 +1,59 @@
-	TOOLCHAIN_DIR := $(subst -none,,$(TRIPLE))-4.9
-endif
+# lowest 64-bit API level
+API_LEVEL := 21
```
This is Android 5.0, Lollipop, which has ~99.6% adoption as per https://apilevels.com/. Additionally, this is from 2014, which is still 10 years ago.

labath commented:
> @labath there have been no additional comments; is there anyone specifically you'd like me to seek feedback from?

No, this is fine. Thanks for reminding me.

compnerd commented:
Going to go ahead and merge this since there/s not been any additional feedback.
